torch_xla/csrc/tensor.cpp

```diff
@@ -294,9 +304,19 @@ void XLATensor::AssignIrValue(torch::lazy::Value ir_value) const {
   ShardingSpecPtr sharding = sharding_spec();
   if (sharding != nullptr) {
     if (!ir_value) {
-      // Create a tensor node if applicable, re-use the current IR otherwise.
-      // TODO(yeounoh) this has some performance implications for convolution.
-      ir_value = GetIrValue();
+      // User provided sharding annotation is accompanied with sharded data
+      // handle. If sharded, we should create a tensor node with the current
+      // data handle instead. If there is no device data (e.g., view), then use
+      // the current IR to preserve the sharding.
+      // TODO(yeounoh) this may not be needed when functionalization is fully
+      // implemented.
+      torch::lazy::BackendDataPtr handle = CurrentDataHandle();
+      if (handle != nullptr) {
+        ir_value = CreateTensorNode(handle, /*read_only=*/false);
+      } else {
+        ir_value = CurrentIrValue();
+      }
+      XLA_CHECK(ir_value);
```

Comment:
hmm what if current `XLATensor` already have a shardingspec and the `ir_value` passed also have a shardingspec, are we going to error out here?

Reply:
The new ir_value as input would not have the sharding spec, unless it's provided by the user via mark_sharding. Also, once the sharding is attached to the tensor and its IR, it cannot be modified unless it's first cleared from both XLATensor and IR.

Reply:
if `ir_value` is empty, doesn't it means we want to reset the IR value to null? If that's the case shouldn't we just leave it as it is. We can attach the sharding_spec when we create the real IR for the current XLATensor.

Reply:
Hm, I was thinknig that the empty IR would be replaced with a tensor node here, since the sharding annotation implies sharded data handle here. WIth the new change in the `CollectSyncTensor` we may be able to simplify (skip) here, let me test though.

Reply:
Ok for `AssignIrValue` with an empty node, we can skip and rely on `CollectSyncTensor`.